aten/src/ATen/native/Copy.cpp

```diff
@@ -242,21 +197,88 @@ static Tensor & copy_impl(Tensor & self, const Tensor & src, bool non_blocking)
     device_type = kMPS;
   }
 
+  // XXX: Move to backend-specific code below?
   // TODO: if we need to, we can also enable this path for quantized tensor
   if (device_type == kCPU && copy_transpose_valid(self, src) && !self.is_quantized()) {
     copy_same_type_transpose_(self, src);
     return self;
   }
 
+  if (!self.is_complex() && src.is_complex()) {
+    TORCH_WARN_ONCE("Casting complex values to real discards the imaginary part");
+  }
+
+  // Check here that we have enough numel in self, but use src numel when
+  // copying to not go out of bounds (see fbgemm::Float16ToFloat_ref). Make sure
+  // this is only enforced right before the backend-specific code is called.
+  TORCH_CHECK(
+    self.numel() >= src.numel(),
+    "Expected self to have more or equal number of elements than src, but got "
+    "self: ", self.numel(), " and src: ", src.numel());
+
+  // Unless there's a good reason not to do it, place backend-specific code here
+  // to make sure that behavior is consistent between backends. There's a number
+  // of early exits above. Also, TensorIteratorConfig().build() checks for
+  // broadcasting via mark_resize_outputs().
```

Comment:
What if different backends already implement these checks in their specific cases and they don't want to incur on the penalty of having to build `TensorIterator`?

Reply:
After discussing offline, I went with a smaller hotfix. I only check numel to avoid reading OOB here:
```cpp
void Float16ToFloat_ref(const float16* src, float* dst, size_t size) {
  for (size_t i = 0; i < size; i++) {
    dst[i] = cpu_half2float(src[i]);
  }
}
```

I added testcases to highlight differences in behavior between fbgemm code and TI.